modin/experimental/xgboost/xgboost.py

```diff
@@ -0,0 +1,499 @@
+            (start_idx + row_parts_last_idx)
+            if (start_idx + row_parts_last_idx < len(row_partitions))
+            else len(row_partitions)
+        )
+
+        actor_row_partitions = []
+        order_of_indexes = []
+        while first < last:
+            actor_row_partitions.append(row_partitions[first][1])
+            order_of_indexes.append(first)
+            first += 1
+        start_idx += row_parts_last_idx
+
+        row_partitions_by_actors.append((ip, order_of_indexes, actor_row_partitions))
+
+    return row_partitions_by_actors
+
+
+def train(
+    params: Dict,
+    dtrain: ModinDMatrix,
+    *args,
+    evals=(),
+    nthread: Optional[int] = cpu_count(),
+    evenly_data_distribution: Optional[bool] = True,
+    **kwargs,
+):
+    """
+    Train XGBoost model.
+
+    Parameters
+    ----------
+    params : dict
+        Booster params.
+    dtrain : ModinDMatrix
+        Data to be trained.
+    nthread: int
+        Number of threads for using in each node. By default it is equal to
+        number of threads on master node.
+    evenly_data_distribution: boolean, default True
+        Whether make evenly distribution of partitions between nodes or not.
+        In case `False` minimal datatransfer between nodes will be provided
+        but the data may not be evenly distributed.
+    \\*\\*kwargs:
+        Other parameters are the same as `xgboost.train` except for
+        `evals_result`, which is returned as part of function return value
+        instead of argument.
+
+    Returns
+    -------
+    dict
+        A dictionary containing trained booster and evaluation history.
+        `history` field is the same as `eval_result` from `xgboost.train`.
+
+        .. code-block:: python
+
+            {'booster': xgboost.Booster,
+             'history': {'train': {'logloss': ['0.48253', '0.35953']},
+                         'eval': {'logloss': ['0.480385', '0.357756']}}}
+    """
+    LOGGER.info("Training started")
+
+    s = time.time()
+
+    X, y = dtrain
+    assert len(X) == len(y)
+    X_row_parts = unwrap_row_partitions(X, bind_ip=not evenly_data_distribution)
+    y_row_parts = unwrap_row_partitions(y, bind_ip=not evenly_data_distribution)
+    assert len(X_row_parts) == len(y_row_parts), "Unaligned train data"
+
+    # Create remote actors
+    actors = create_actors(nthread=nthread)
+
+    add_as_eval_method = None
+    if len(evals):
+        for (eval_data, method) in evals:
+            if id(eval_data) == id(dtrain):
+                add_as_eval_method = method
+                evals.remove((eval_data, method))
+
+        evals_unwrapped = [
+            (
+                (
+                    unwrap_row_partitions(eval_X, bind_ip=not evenly_data_distribution),
+                    unwrap_row_partitions(eval_y, bind_ip=not evenly_data_distribution),
+                    eval_method,
+                )
+            )
+            for ((eval_X, eval_y), eval_method) in evals
+        ]
```

Comment:
why do this here instead of explicitly calling `unwrap_row_partitions(eval_X)` for arguments of `_split_data_across_actors`? seems like two loops where one is enough

Reply:
Thanks, explicitly calling was added for argoments of `_split_data_across_actors`.